### ISSUE TITLE
Feature: implement caching mechanism

### ISSUE THREAD

**Issue by lwansbrough** [balderdashy/waterline#200]
Hey @particlebanana et al, I'm thinking it would be very useful to be able to provide a mechanism for specifying caching methods. Personally I think it would be very useful to have this functionality available within Waterline due to the importance of the way in which a cached object's key is generated. I was considering doing this outside of Waterline - but then I wouldn't have access to which properties were being populated, etc in order to build a corresponding key. Here's what I'm thinking:

Specifying a cache for the query would look something like this:
    
    User.findOne(123).populate('roles').cache({
        get: function(key) {
            // call to memcached/redis/etc here
        },
        set: function(key, value) {
            // call to memcached/redis/etc here
        }
    }).exec(function(err, user) { console.log(user); }));

The key would be a hashed representation of the normalized criteria/joins/options etc and value of course being the model. `remove` and `update` could also be specified for `User.update` and `User.remove` queries.

I will probably end up doing this anyway, but I want to request for comments in case you've got a better solution for this.

Cheers

**Comment by devinivy** [comment]
Been thinking about it.  Here's my brain-dump on the topic.

I think this project would be high-value for the amount of time it will take to implement.  I'd suggest starting by implementing a write-around cache, probably at the collection level.  Junction tables can inherit caching from their dominant collection.  One option is to create the idea of a "cache adapter" which implements certain methods that enable caching when composed with another adapter, i.e. taking advantage of the already-existing-and-sort-of-working syntax `adapter: ['mongo', 'redis']` in collection definition.  One last idea is to let the adapter for the persistent data store optionally determine the key used in the cache.

A look at sequelize's solution (in-progress):
They're going with a more general "plugin" route, which seem to be an extension of their hooks: https://github.com/sequelize/sequelize/issues/803
Here's the redis cache layer for sequelize: https://github.com/rfink/sequelize-redis-cache
It only hooks into reads, so it's write-around caching.

Perhaps we could use https://www.npmjs.com/package/json-stable-stringify to help create a deterministic hash of `find` options, and cache using that as a key.

**Comment by tjwebb** [comment]
I'd be open to using redis as an optional caching layer: you should take a
look at waterline-cursor which is the module I think would benefit most
from this.

**Comment by devinivy** [comment]
Where do you see waterline-cursor benefiting from this?  Doesn't that module do the in-memory joins?  (Perhaps I'm misunderstanding something!)  I'm thinking about query/result caching.  I guess it could benefit waterline-cursor in the sense that waterline-cursor may not have to be used as often!

**Comment by tjwebb** [comment]
Yea, e.g. what if you wanted to do a join via waterline-cursor of
non-trivial size? Be nice not to rely only on ram.

I like the general idea of a key/value adapter spec.

**Comment by stuk88** [comment]
User.findOne(123).populate('roles').cache(60000).exec(function(err, user) { console.log(user); })); 

would be awesome

**Comment by ronenteva** [comment]
+1 for @stuk88 idea

**Comment by vidia** [comment]
+1 as well to adding a cache.

**Comment by bojidar-bg** [comment]
Though it won't help much in pushing things forward, but still.... :+1:

**Comment by sailsbot** [comment]
Thanks for posting, @lwansbrough.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sailsbot** [comment]
Thanks for posting, @lwansbrough.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by luislobo** [comment]
Can anyone reopen this one? Its a Feature issue... I guess Feature issues should never be auto closed?

**Comment by Benew** [comment]
+1 please reopen

**Comment by particlebanana** [comment]
This is linked to on the [Roadmap](https://github.com/balderdashy/waterline/blob/master/ROADMAP.md). We are trying to keep issues as only issues and track features using the Roadmap. Closing this doesn't mean it's no longer valid. Caching is very cool and something we will keep in mind as features progress.

**Comment by atiertant** [comment]
this is now available in [offshore 0.0.4](https://github.com/Atlantis-Software/offshore), documentation could be found [here](https://github.com/Atlantis-Software/offshore-docs/blob/master/queries/query.md#cache)
